Godot XR Update May 2026 (#1333)
godotengine/godot-website: collections/_article/godot-xr-update-may-2026.md

---
title: "Godot OpenXR Vendors Plugin v5.1"
excerpt: "Introducing the new major version of the Godot OpenXR Vendors plugin with several Android XR features!"
categories: ["progress-report"]
author: Bastiaan Olij, David Snopek, and Fredia Huya-Kouadio
image: /storage/blog/covers/godot-openxr-vendors-plugin-v5-1-godot-xr-community.jpg
image_caption_description: "Expedition to Blobotopia - David Snopek and Logan Lang"
date: 2026-05-19 12:00:00
---

> ## Godot OpenXR Vendors Plugin v5.1

Reviewer: ```suggestion
## Godot OpenXR Vendors plugin v5.1
```

However, enabling them on standalone Android-based headsets is much trickier. You'd need to build the layers yourself, and manually add them to your Godot project at a specific location, along with a special JSON file.

Version 5 of the Godot OpenXR Vendors plugin simplifies this process by building the Android binaries for the OpenXR validation layers, and wrapping them into an Android library binary (AAR). For added convenience, the generated Android library binary (AAR) is uploaded to [MavenCentral](https://central.sonatype.com/artifact/org.godotengine/openxr-validation-layers/overview) alongside each release of the Godot OpenXR Vendors plugin. 
This enables OpenXR Android projects to easily integrate the validation layers by adding a dependency to the [OpenXR validation layers Android library](https://godotvr.github.io/godot_openxr_vendors/manual/openxr_validation_layers.html).

For Godot projects built with the Godot OpenXR Vendors plugin v5, all you need to do is click a checkbox in your export settings!

> <img src="/storage/blog/godot-xr/updates/validation_layer_selection.png" alt="Enable OpenXR Validation Layers"/>

Reviewer: ```suggestion
<img src="/storage/blog/godot-xr/updates/validation_layer_selection.png" alt="Enable OpenXR Validation Layers checkbox in export settings"/>
```

---
title: "Godot XR update — May 2026"

> excerpt: "May update from the Godot XR Team, new vendor plugin features!"

Reviewer 2: This post is focused on the vendors plugin, so let's update the title and excerpt to match.
```suggestion
title: "Godot OpenXR Vendors Plugin v5.1"
excerpt: "Introducing the new major version of the Godot OpenXR Vendors plugin with several Android XR features!"
```

categories: ["progress-report"]

> author: Bastiaan Olij, David Snopek and Fredia Huya-Kouadio

Reviewer: ```suggestion
author: Bastiaan Olij, David Snopek, and Fredia Huya-Kouadio
```

image: /storage/blog/covers/may-2026-update-godot-xr-community.jpg
image_caption_description: "TBD"
date: 2026-05-19 12:00:00
---

> ## Godot OpenXR Vendors Plugin v5

Reviewer 2: ```suggestion
## Godot OpenXR Vendors Plugin v5.1
```

> The [OpenXR Vendors plugin](https://github.com/GodotVR/godot_openxr_vendors) is an official GDExtension that builds on top of the OpenXR support in Godot, and provides support for additional functionality, primarily, vendor-specific OpenXR extensions.

Reviewer: ```suggestion
The [OpenXR Vendors plugin](https://github.com/GodotVR/godot_openxr_vendors) is an official GDExtension that builds on top of the OpenXR support in Godot, and provides support for additional functionality, primarily vendor-specific OpenXR extensions.
```

> We recently released the next major version (v5) and one minor update, 5.1.0-stable, which can be downloaded from the Asset Store, Asset Library or GitHub release.

Reviewer: ```suggestion
We recently released the next major version (v5) and one minor update, 5.1.0-stable, which can be downloaded from the Asset Store, Asset Library, or GitHub release page.
```

> We recently released the next major version (v5) and one minor update, 5.1.0-stable, which can be downloaded from the Asset Store, Asset Library or GitHub release.

Reviewer 2: ```suggestion
Today we are releasing version 5.1, which can be downloaded from the Asset Store, Asset Library or GitHub release.
```

We'll need to add links for the various download locations.

Author: I've added the links in. Obviously the 5.1.0 release link isn't valid yet until we do the actual build.

> ### Update and Refactor

Reviewer: ```suggestion
### Update and refactor
```

This is due in large part to the collaboration between Google, the Godot Foundation, and W4 Games which was [announced in November 2025](https://godotengine.org/article/announcing-android-xr-support/).

In v5.0, that included two new features (both of which depended on the Godot 4.6 upgrade):

- [Dynamic Resolution](https://registry.khronos.org/OpenXR/specs/1.1/html/xrspec.html#XR_ANDROID_recommended_resolution): Will dynamically scale down the render resolution in response to system load, which can help maintain a consistent frame rate. We implemented a [similar feature](https://godotengine.org/article/godot-openxr-vendors-plugin-400/#dynamic-resolution) for Meta headsets back in v4.
- [Unbounded Reference Space](https://registry.khronos.org/OpenXR/specs/1.1/html/xrspec.html#XR_ANDROID_unbounded_reference_space): Enables the user to move freely through a complex environment, even many meters from where they started.

In v5.1, there's additionally:

> - [Trackables](https://godotvr.github.io/godot_openxr_vendors/manual/androidxr/trackables.html): Provides information about the physical space the user is in. Android XR also supports [OpenXR spatial entities](https://docs.godotengine.org/en/latest/tutorials/xr/openxr_spatial_entities.html), which provides similar functionality, but at the moment there's still features in the Android XR-specific trackables, that aren't supported yet in the multi-vendor spatial entities.

Reviewer: ```suggestion
- [Trackables](https://godotvr.github.io/godot_openxr_vendors/manual/androidxr/trackables.html): Provides information about the physical space the user is in. Android XR also supports [OpenXR spatial entities](https://docs.godotengine.org/en/latest/tutorials/xr/openxr_spatial_entities.html), which provides similar functionality, but at the moment there's still features in the Android XR-specific trackables that aren't supported yet in the multi-vendor spatial entities.
```

- **Automatically launching the streaming client**: Using Android XR's Direct Preview functionality, you can run XR applications on your desktop which are streamed to the headset, which can allow for faster iteration time. Now, the plugin can make this easier by automatically launching the streaming client on your headset.
- **Mouse interaction**: Allows accessing mouse input in a similar way to traditional XR controllers.

> ### OpenXR Validation Layers for Android

Reviewer: ```suggestion
### OpenXR validation layers for Android
```

> The Khronos Group provides validation layers for OpenXR that can perform extra validation to ensure that Godot is using OpenXR correctly. Enabling these on desktop is very easy, and Khronos even provides pre-built binaries for Windows and MacOS.

Reviewer: ```suggestion
The Khronos Group provides validation layers for OpenXR that can perform extra validation to ensure that Godot is using OpenXR correctly. Enabling these on desktop is very easy, and Khronos even provides pre-built binaries for Windows and macOS.
```

> However, enabling them on standalone Android-based headsets is much trickier! You'd need to build the layers yourself, and manually add them to your Godot project at a specific location, along with a special JSON file.

Reviewer: ```suggestion
However, enabling them on standalone Android-based headsets is much trickier. You'd need to build the layers yourself, and manually add them to your Godot project at a specific location, along with a special JSON file.
```

Version 5 of the Godot OpenXR Vendors plugin simplifies this process by building the Android binaries for the OpenXR validation layers, and wrapping them into an Android library binary (AAR). For added convenience, the generated Android library binary (AAR) is uploaded to [MavenCentral](https://central.sonatype.com/artifact/org.godotengine/openxr-validation-layers/overview) alongside each release of the Godot OpenXR Vendors plugin. 

> This enables OpenXR Android projects to easily integrate the validation layers by adding a dependency to the [OpenXR validation layers Android library on MavenCentral](https://central.sonatype.com/artifact/org.godotengine/openxr-validation-layers/overview).

Reviewer 3: Do we need to link https://central.sonatype.com/artifact/org.godotengine/openxr-validation-layers/overview twice?

And, actually, should we link this page somewhere in here? https://godotvr.github.io/godot_openxr_vendors/manual/openxr_validation_layers.html

cc @m4gr3d who added the MavenCentral bits

The main 3D content that's rendered by Godot is also associated with a composition layer, the main projection layer. However, up until now, it wasn't possible to configure any of these extensions on that layer. That's changed in v5.1!

Now you can set any of the same extra configuration that was available via **OpenXRCompositionLayer** nodes on the main projection layer too, either through code or Project Settings.

On some headsets (namely, Meta's), this can allow using the `XR_FB_composition_layer_depth_test` extension to allow content rendered by Godot to correctly appear in front of or behind other composition layers, without having to use hole punching (which can't be used with transparent layers). This is a feature that has been requested by the community for a long time.

### Hybrid App testing on desktop

In v4, we added support for Hybrid Apps on Android, which can dynamically switch between fully immersive and floating 2D panels.

However, developers couldn't test them on desktop, only when deploying to the headset.

In v5, we've added the ability to simulate Hybrid App switching on desktop, which can help to accelerate iteration speed.

-----

> *The illustration picture for this article comes from* ...

Reviewer 3: If we want to use a screenshot from Expedition to Blobotopia, here's one:

<img width="3840" height="2160" alt="Image" src="https://github.com/user-attachments/assets/2f07de06-e79b-4ada-bcd0-3192fc96eee6" />

Potential description (sorry if I crammed too many links in there :-)):

```suggestion
*The illustration picture for this article comes from [Expedition to Blobotopia](https://dsnopek.itch.io/expedition-to-blobotopia) which was [created entirely in VR](https://www.youtube.com/watch?v=6RE8KuCspqw) and available for [Meta](https://www.meta.com/experiences/expedition-to-blobotopia-demo/24941992168770375/) and [Android XR](https://play.google.com/store/apps/details?id=com.snopekgames.gwj81) headsets.*
```
